docs/configure.md
# Configure

## Resolution

Camera capture and resolution exposed is controlled by three parameters:

- `--camera-width` and `--camera-height` define the camera capture resolution
- `--camera-snapshot.height` - define height for an aspect ratio scaled resolution for `/snapshot` (JPEG) output - this might require rescaller and might not always work
- `--camera-video.height` - define height for an aspect ratio scaled resolution for `/video` and `/webrtc` (H264) output - this might require rescaller and might not always work, this is no larger than `--camera-snapshot.height`
- `--camera-stream.height` - define height for an aspect ratio scaled resolution for `/stream` (MJPEG) output - this might require rescaller and might not always work, this is no larger than `--camera-video.height`

The resolution scaling is following the `snapshot >= video >= stream`:

- `snapshot` provides a high quality image for the purpose of timelapses (3Mbps-10Mbps)
- `video` is an efficient high-quality H264 video stream (3-10Mbps)
- `stream` is an inefficient MJPEG stream requiring significant amount of bandwidth (10-100Mbps)

### Example: Raspberry PI v3 Camera (best)

```text
libcamera-still --list-cameras
Available cameras
-----------------
0 : imx708_wide [4608x2592] (/base/soc/i2c0mux/i2c@1/imx708@1a)
    Modes: 'SRGGB10_CSI2P' : 1536x864 [120.13 fps - (0, 0)/4608x2592 crop]
                             2304x1296 [56.03 fps - (0, 0)/4608x2592 crop]
                             4608x2592 [14.35 fps - (0, 0)/4608x2592 crop]
```

To get the best resolution for the above camera, the `--camera-width` and `--camera-height`
needs to be configured with the native resolution of camera sensor, only then the resolution
can be downscaled.

```text
--camera-width=2304 --camera-height=1296 --camera-snapshot.height=1080 --camera-video.height=720 --camera-stream.height=480
```

This will result in:

- camera will capture `2304x1296` a full sensor
- `snapshot` be ~1920x1080
- `video` be ~1280x720
- `stream` be ~640x480

### Example: Raspberry PI v3 Camera with cropped output (bad)

If the camera capture is not configured properly the result image will be cropped.

```text
--camera-width=1920 --camera-height=1080 --camera-snapshot.height=1080 --camera-video.height=720 --camera-stream.height=480
```

This will result in:

- camera will capture the middle `1920x1080` from the `2304x1296`
- `snapshot` be ~1920x1080
- `video` be ~1280x720
- `stream` be ~640x480

## List all available controls

You can view all available configuration parameters by adding `--log-verbose`
to one of the above commands, like:

```bash
tools/libcamera_camera.sh --camera-list_options ...
```

Or navigate to `http://<ip>:8080/option` while running.

Depending on control they have to be used for camera, ISP, or JPEG or H264 codec:

```bash
# specify camera option
--camera-options=brightness=1000

# specify ISP option
--camera-isp.options=digital_gain=1000

# specify H264 option
--camera-video.options=bitrate=10000000

# specify snapshot option
--camera-snapshot.options=compression_quality=60
--camera-stream.options=compression_quality=60
```

## List all available formats and use proper one

You might list all available capture formats for your camera:

```bash
v4l2-ctl -d /dev/video0 --list-formats-ext
```

Some of them might be specified to streamer:

```bash
tools/*_camera.sh --camera-format=RG10 # Bayer 10 packed
tools/*_camera.sh --camera-format=YUYV
tools/*_camera.sh --camera-format=MJPEG
tools/*_camera.sh --camera-format=H264 # This is unstable due to h264 key frames support
```

It is advised to always use:

- for `libcamera` the `--camera-type=libcamera --camera-format=YUYV` (better image quality) or `--camera-format=YUV420` (better performance)
- for `USB cameras` the `--camera-type=libcamera --camera-format=MJPEG`


debian/changelog
camera-streamer (0.1.0) unstable; urgency=medium

  * Initial Release.

 -- Kamil Trzciński <cs@ayufan.eu>  Fri, 26 May 2023 13:15:22 +0200


cmd/camera-streamer/opts.c
#include "util/http/http.h"
#include "util/opts/opts.h"
#include "util/opts/log.h"
#include "util/opts/fourcc.h"
#include "device/camera/camera.h"
#include "output/rtsp/rtsp.h"
#include "output/webrtc/webrtc.h"
#include "output/output.h"

camera_options_t camera_options = {
  .path = "",
  .width = 1920,
  .height = 1080,
  .format = 0,
  .nbufs = 3,
  .fps = 30,
  .allow_dma = true,
  .high_res_factor = 0.0,
  .low_res_factor = 0.0,
  .auto_reconnect = 0,
  .auto_focus = true,
  .options = "",
  .list_options = false,
  .snapshot = {
    .options = "compression_quality=80"
  },
  .stream = {
    .options = "compression_quality=80"
  },
  .video = {
#ifdef USE_HW_H264
    .disabled = 0,
#else // USE_HW_H264
    .disabled = 1,
#endif
    .options =
      "video_bitrate_mode=0" OPTION_VALUE_LIST_SEP
      "video_bitrate=2000000" OPTION_VALUE_LIST_SEP
      "repeat_sequence_header=5000000" OPTION_VALUE_LIST_SEP
      "h264_i_frame_period=30" OPTION_VALUE_LIST_SEP
      "h264_level=11" OPTION_VALUE_LIST_SEP
      "h264_profile=4" OPTION_VALUE_LIST_SEP
      "h264_minimum_qp_value=16" OPTION_VALUE_LIST_SEP
      "h264_maximum_qp_value=32"
  }
};

http_server_options_t http_options = {
  .port = 8080,
  .maxcons = 10
};

log_options_t log_options = {
  .debug = false,
  .verbose = false
};

rtsp_options_t rtsp_options = {
  .port = 0,
  .allow_truncated = false
};

webrtc_options_t webrtc_options = {
};

option_value_t camera_formats[] = {
  { "DEFAULT", 0 },
  { "YUYV", V4L2_PIX_FMT_YUYV },
  { "YUV420", V4L2_PIX_FMT_YUV420 },
  { "YUYV", V4L2_PIX_FMT_YUYV },
  { "NV12", V4L2_PIX_FMT_NV12 },
  { "NV21", V4L2_PIX_FMT_NV21 },
  { "MJPG", V4L2_PIX_FMT_MJPEG },
  { "MJPEG", V4L2_PIX_FMT_MJPEG },
  { "JPEG", V4L2_PIX_FMT_MJPEG },
  { "H264", V4L2_PIX_FMT_H264 },
  { "RG10", V4L2_PIX_FMT_SRGGB10 },
  { "GB10P", V4L2_PIX_FMT_SGRBG10P },
  { "RG10P", V4L2_PIX_FMT_SRGGB10P },
  { "BG10P", V4L2_PIX_FMT_SBGGR10P },
  { "RGB565", V4L2_PIX_FMT_RGB565 },
  { "RGBP", V4L2_PIX_FMT_RGB565 },
  { "RGB24", V4L2_PIX_FMT_RGB24 },
  { "RGB", V4L2_PIX_FMT_RGB24 },
  { "BGR", V4L2_PIX_FMT_BGR24 },
  {}
};

option_value_t camera_type[] = {
  { "v4l2", CAMERA_V4L2 },
  { "libcamera", CAMERA_LIBCAMERA },
  { "dummy", CAMERA_DUMMY },
  {}
};

option_t all_options[] = {
  DEFINE_OPTION_PTR(camera, path, string, "Chooses the camera to use. If empty connect to default."),
  DEFINE_OPTION_VALUES(camera, type, camera_type, "Select camera type."),
  DEFINE_OPTION(camera, width, uint, "Set the camera capture width."),
  DEFINE_OPTION(camera, height, uint, "Set the camera capture height."),
  DEFINE_OPTION_VALUES(camera, format, camera_formats, "Set the camera capture format."),
  DEFINE_OPTION(camera, nbufs, uint, "Set number of capture buffers. Preferred 2 or 3."),
  DEFINE_OPTION(camera, fps, uint, "Set the desired capture framerate."),
  DEFINE_OPTION_DEFAULT(camera, allow_dma, bool, "1", "Prefer to use DMA access to reduce memory copy."),
  DEFINE_OPTION(camera, high_res_factor, float, "Set the desired high resolution output scale factor."),
  DEFINE_OPTION(camera, low_res_factor, float, "Set the desired low resolution output scale factor."),
  DEFINE_OPTION_PTR(camera, options, list, "Set the camera options. List all available options with `-camera-list_options`."),
  DEFINE_OPTION(camera, auto_reconnect, uint, "Set the camera auto-reconnect delay in seconds."),
  DEFINE_OPTION_DEFAULT(camera, auto_focus, bool, "1", "Do auto-focus on start-up (does not work with all camera)."),
  DEFINE_OPTION_DEFAULT(camera, force_active, bool, "1", "Force camera to be always active."),
  DEFINE_OPTION_DEFAULT(camera, vflip, bool, "1", "Do vertical image flip (does not work with all camera)."),
  DEFINE_OPTION_DEFAULT(camera, hflip, bool, "1", "Do horizontal image flip (does not work with all camera)."),

  DEFINE_OPTION_PTR(camera, isp.options, list, "Set the ISP processing options. List all available options with `-camera-list_options`."),

  DEFINE_OPTION_PTR(camera, snapshot.options, list, "Set the JPEG compression options. List all available options with `-camera-list_options`."),
  DEFINE_OPTION(camera, snapshot.height, uint, "Override the snapshot height and maintain aspect ratio."),

  DEFINE_OPTION_DEFAULT(camera, stream.disabled, bool, "1", "Disable stream."),
  DEFINE_OPTION_PTR(camera, stream.options, list, "Set the JPEG compression options. List all available options with `-camera-list_options`."),
  DEFINE_OPTION(camera, stream.height, uint, "Override the stream height and maintain aspect ratio."),

  DEFINE_OPTION_DEFAULT(camera, video.disabled, bool, "1", "Disable video."),
  DEFINE_OPTION_PTR(camera, video.options, list, "Set the H264 encoding options. List all available options with `-camera-list_options`."),
  DEFINE_OPTION(camera, video.height, uint, "Override the video height and maintain aspect ratio."),

  DEFINE_OPTION_DEFAULT(camera, list_options, bool, "1", "List all available options and exit."),

  DEFINE_OPTION(http, port, uint, "Set the HTTP web-server port."),
  DEFINE_OPTION(http, maxcons, uint, "Set maximum number of concurrent HTTP connections."),

  DEFINE_OPTION_DEFAULT(rtsp, port, uint, "8554", "Set the RTSP server port (default: 8854)."),

  DEFINE_OPTION_PTR(webrtc, ice_servers, list, "Specify ICE servers: [(stun|turn|turns)(:|://)][username:password@]hostname[:port][?transport=udp|tcp|tls)]."),
  DEFINE_OPTION_DEFAULT(webrtc, disable_client_ice, bool, "1", "Ignore ICE servers provided in '/webrtc' request."),

  DEFINE_OPTION_DEFAULT(log, debug, bool, "1", "Enable debug logging."),
  DEFINE_OPTION_DEFAULT(log, verbose, bool, "1", "Enable verbose logging."),
  DEFINE_OPTION_DEFAULT(log, stats, uint, "1", "Print statistics every duration."),
  DEFINE_OPTION_PTR(log, filter, list, "Enable debug logging from the given files. Ex.: `-log-filter=buffer.cc`"),

  {}
};
